Allow using exported types in component and instance types (#102)
WebAssembly/component-model: design/mvp/Explainer.md

componentdecl ::= <importdecl>
                | <instancedecl>
instancedecl  ::= core-prefix(<core:type>)

>                 | (type <typebound>)

Reviewer: ```suggestion
                | <typedecl>
```
?  Otherwise, I don't see where `typedecl` is used

                | (core module <typeidx>)
                | (func <typeidx>)
                | (value <typeidx>)
                | (instance <typeidx>)

>                 | (component <typeidx>)

Reviewer: If we add the syntactic sugar suggested below, I think perhaps we can hold off introducing these syntactic forms until they might actually be needed in the future.

                | <alias>
                | <exportdecl>
importdecl    ::= (import <name> bind-id(<importdesc>))

> typedecl      ::= (type <id>? <typebound>)

Reviewer: Interesting; so to see if I'm reading correctly: do *all* type definitions inside a component-/instance-type happen inside a type bound?

> exportdecl    ::= (export <name> <sortidx>)

Reviewer: I think it would be useful to keep allowing the same syntax as now by having:
```
exportdecl ::= (export <name> <exportdesc>)
exportdesc ::= bind-id(<externdesc>)
             | <sortidx>
```

importdesc    ::= (<sort> (type <u32>) )
                | core-prefix(<core:moduletype>)
                | <functype>
                | <componenttype>
                | <instancetype>
                | (value <valtype>)
                | (type <typebound>)

> typebound     ::= (eq <defype>)

Reviewer: Should `typebound` continue to also allow `typeidx`?

present only in component types).

As with core modules, `importdecl` and `exportdecl` classify component
`import` and `export` definitions, with `importdecl` allowing an
identifier to be bound for use within the type. The definition of
`externname` is given in the [imports and
exports](#import-and-export-definitions) section below. Following the
precedent of [`core:typeuse`], the text format allows both references
to out-of-line type definitions (via `(type <typeidx>)`) and inline
type expressions that the text format desugars into out-of-line type
definitions.

The `type` and `core:type` declarators simply introduce a simple type
alias into the local type index space. These entries are fully
transparent to consumers of the component type, and any later
references to theme could (semantically, if not syntactically) be
replaced by the definition provided, as in fact they are during
validation elaboration.

Export declarators take a name and an extern descriptor describing the
entry that they are exporting. However, they also create an entry in a
local (to the current module or instance type declaration) index space

> for the appropriate sort, which may be used by later declarators to
> refer to the concept of this export. In the future, this will be used

Reviewer: ```suggestion
for the appropriate sort, which may be used by later declarators to
refer to the abstract named thing being exported. In the future, this will be used
```

to export bounded resource types as part of the addition of [resource
and handle types]. In the short term, allowing the export of existing
non-resource types (always with an equality type-bound) has two main
use cases:
* Type exports allow a component or interface to associate a name with a
  structural type (e.g., `(export "nanos" (type (eq u64)))`) which bindings
  generators can use to generate type aliases (e.g., `typedef uint64_t nanos;`).
* Type imports and exports can provide additional information to toolchains and
  runtimes for defining the behavior of host APIs.

Note that both `(type ...)` and `({im,ex}port ... (type ...))`

> declarators introduce entries to the type index space. The difference
> is that the former are treated completely as aliases, with no identity
> of their own, while the latter are intended to associate future
> declarators with the exported type, so that the latter is correctly
> used in WIT/bindings/etc. When resource types are added, only the
> latter form will allow adding abstract resource types to the index
> space.

Reviewer: ```suggestion
declarators introduce entries to the type index space. Before resource
types, the only difference is that the latter are associated with a name.
(With resource types, only type imports/exports will be able to introduce
introduce resource types into the type index space since resource types.)
```
I think "alias" might be a confusing distinction to make here since, from another point of view, type exports are more like aliases (they add another index for the same thing, rather than defining a totally new thing).  This suggested alternative wording tries to avoid the confusion.

As an example, consider:
``` wasm
(component $C
  (type $D (component
    (type $u u32)
    (export "nanos" (type $n (eq u32)))
    (export "foo" (func (param "in" $u) (result $n))))
```

While both types are allocated in the same index space, they are
treated differently in e.g. WIT:
```wit
type nanos = u32
foo: func(in: u32) -> nanos
```

> Component import and alias declarators are analagous to the core module
> declarators introduced above, and have essentially the same meaning, although
> they also (like the import and alias _definitions_ present inside a component)
> create entries in the local index spaces that may later be exported. As with
> core modules, `importdecl` classifies a component's `import` definitions.

Reviewer: ```suggestion
Component alias declarators are analagous to the core module
declarators introduced above, and have essentially the same meaning.
```
I think the parts I chopped out are already said above?
